File: scripts/zigux/check-phase8-scripts-readme-broader-packet.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import tempfile
from dataclasses import dataclass
from pathlib import Path


def _default_root() -> Path:
    resolved = Path(__file__).resolve()
    if len(resolved.parents) >= 3:
        return resolved.parents[2]
    return resolved.parent


ROOT = _default_root()
SCRIPTS_README = Path("scripts/zigux/README.md")
PHASE8_VALIDATOR = Path("scripts/zigux/validate-phase8.py")
HELP_KALLSYMS_PACKET_CHECKER = Path("scripts/zigux/check-phase8-help-kallsyms-packet.py")
LIBBPF_SHARD_ROUTES_CHECKER = Path("scripts/zigux/check-phase8-libbpf-shard-routes.py")
TESTS_README = Path("zigux/tests/README.md")

REQUIRED_FILES = (
    SCRIPTS_README,
    PHASE8_VALIDATOR,
    HELP_KALLSYMS_PACKET_CHECKER,
    LIBBPF_SHARD_ROUTES_CHECKER,
    TESTS_README,
)

FILE_MARKERS: dict[Path, tuple[str, ...]] = {
    SCRIPTS_README: (
        "## Phase 8",
        "`tools/lib/subcmd/help.zig`",
        "`tools/lib/symbol/kallsyms.zig`",
        "`zigux/tests/phase8_help.zig`",
        "`zigux/tests/phase8_kallsyms.zig`",
        "`zigux/tests/phase8_help_only_build.zig`",
        "`zigux/tests/phase8_help_kallsyms_only_build.zig`",
        "`tools/lib/bpf/zigux_segments/verify.zig`",
        "`zigux/tests/phase8_libbpf_segments.zig`",
        "`zigux/tests/phase8_libbpf_segments_only_build.zig`",
        "current public-tree rereads plus the shared packet guards `scripts/zigux/check-phase8-help-kallsyms-packet.py` and `scripts/zigux/check-phase8-libbpf-shard-routes.py` rematerialize those broader help, kallsyms, and libbpf-segment companions on `master`",
    ),
    PHASE8_VALIDATOR: (
        'HELP_KALLSYMS_PACKET_CHECKER = Path("scripts/zigux/check-phase8-help-kallsyms-packet.py")',
        'LIBBPF_SHARD_ROUTES_CHECKER = Path("scripts/zigux/check-phase8-libbpf-shard-routes.py")',
        "HELP_KALLSYMS_PACKET_CHECKER,",
        "LIBBPF_SHARD_ROUTES_CHECKER,",
        'Path("scripts/zigux/README.md"): (',
    ),
    HELP_KALLSYMS_PACKET_CHECKER: (
        "`scripts/zigux/check-phase8-help-kallsyms-packet.py` and `scripts/zigux/check-phase8-libbpf-shard-routes.py` rematerialize those broader help, kallsyms, and libbpf-segment companions on `master`",
        "`tools/lib/subcmd/help.zig`",
        "`tools/lib/symbol/kallsyms.zig`",
        "`zigux/tests/phase8_help_kallsyms_only_build.zig`",
    ),
    LIBBPF_SHARD_ROUTES_CHECKER: (
        'VALIDATOR_PATH = "scripts/zigux/validate-phase8.py"',
        'LIBBPF_SEGMENTS_TEST_PATH = "zigux/tests/phase8_libbpf_segments.zig"',
        'LIBBPF_SEGMENTS_BUILD_PATH = "zigux/tests/phase8_libbpf_segments_only_build.zig"',
        'VERIFY_PATH = "tools/lib/bpf/zigux_segments/verify.zig"',
    ),
    TESTS_README: (
        "`tools/lib/bpf/zigux_segments/verify.zig`",
        "`zigux/tests/phase8_help_kallsyms_only_build.zig`",
        "`zigux/tests/phase8_libbpf_segments.zig`",
        "`zigux/tests/phase8_libbpf_segments_only_build.zig`",
    ),
}


@dataclass
class ValidationResult:
    missing_files: list[str]
    missing_markers: list[str]


def _read(path: Path) -> str:
    return path.read_text(encoding="utf-8")


def _write(path: Path, text: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(text, encoding="utf-8")


def validate_root(root: Path) -> ValidationResult:
    missing_files = [path.as_posix() for path in REQUIRED_FILES if not (root / path).exists()]
    missing_markers: list[str] = []
    for relative_path, markers in FILE_MARKERS.items():
        path = root / relative_path
        if not path.exists():
            continue
        text = _read(path)
        for marker in markers:
            if marker not in text:
                missing_markers.append(f"{relative_path}:{marker}")
    return ValidationResult(missing_files=missing_files, missing_markers=missing_markers)


def emit_result(result: ValidationResult) -> int:
    if result.missing_files or result.missing_markers:
        print("PHASE8_SCRIPTS_README_BROADER_PACKET=fail")
        if result.missing_files:
            print("PHASE8_SCRIPTS_README_BROADER_PACKET_MISSING_FILES_START")
            for item in result.missing_files:
                print(item)
            print("PHASE8_SCRIPTS_README_BROADER_PACKET_MISSING_FILES_END")
        if result.missing_markers:
            print("PHASE8_SCRIPTS_README_BROADER_PACKET_MISSING_MARKERS_START")
            for item in result.missing_markers:
                print(item)
            print("PHASE8_SCRIPTS_README_BROADER_PACKET_MISSING_MARKERS_END")
        return 1

    print("PHASE8_SCRIPTS_README_BROADER_PACKET=pass")
    print(f"PHASE8_SCRIPTS_README_BROADER_PACKET_REQUIRED_FILE_COUNT={len(REQUIRED_FILES)}")
    print(
        "PHASE8_SCRIPTS_README_BROADER_PACKET_REQUIRED_MARKER_COUNT="
        f"{sum(len(markers) for markers in FILE_MARKERS.values())}"
    )
    return 0


def _passing_fixture(root: Path) -> None:
    for relative_path in REQUIRED_FILES:
        if relative_path in FILE_MARKERS:
            _write(root / relative_path, "\n".join(FILE_MARKERS[relative_path]) + "\n")
        else:
            _write(root / relative_path, f"{relative_path.as_posix()}\n")


def run_self_test() -> int:
    case_count = 1
    with tempfile.TemporaryDirectory(prefix="phase8-scripts-readme-broader-packet-") as tmp:
        root = Path(tmp)
        _passing_fixture(root)

        expected_default_root = Path(__file__).resolve()
        if len(expected_default_root.parents) >= 3:
            expected_default_root = expected_default_root.parents[2]
        else:
            expected_default_root = expected_default_root.parent
        if ROOT != expected_default_root:
            raise AssertionError("expected default root to resolve to the repository root")

        passing = validate_root(root)
        if passing.missing_files or passing.missing_markers:
            raise AssertionError("expected passing fixture to validate")

        for relative_path, markers in FILE_MARKERS.items():
            path = root / relative_path
            original = _read(path)
            for marker in markers:
                path.write_text(original.replace(marker, "", 1), encoding="utf-8")
                result = validate_root(root)
                expected = f"{relative_path}:{marker}"
                if expected not in result.missing_markers:
                    raise AssertionError(f"expected missing marker to be reported: {expected}")
                path.write_text(original, encoding="utf-8")
                case_count += 1

        for relative_path in REQUIRED_FILES:
            path = root / relative_path
            original = _read(path)
            path.unlink()
            result = validate_root(root)
            expected = relative_path.as_posix()
            if expected not in result.missing_files:
                raise AssertionError(f"expected missing file to be reported: {expected}")
            _write(path, original)
            case_count += 1

    print("PHASE8_SCRIPTS_README_BROADER_PACKET_SELF_TEST=pass")
    print(f"PHASE8_SCRIPTS_README_BROADER_PACKET_SELF_TEST_CASE_COUNT={case_count}")
    return 0


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--root", type=Path, default=ROOT)
    parser.add_argument("--self-test", action="store_true")
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    if args.self_test:
        return run_self_test()
    return emit_result(validate_root(args.root))


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/zigux/check_phase8_scripts_readme_broader_packet.zig
// Ported from check-phase8-scripts-readme-broader-packet.py by gen_marker_guard.py
const std = @import("std");
const Io = std.Io;
const guard = @import("zigux_guard.zig");

const PASS_MARKER = "PHASE8_SCRIPTS_README_BROADER_PACKET=pass";
const SELF_TEST_PASS_MARKER = "PHASE8_SCRIPTS_README_BROADER_PACKET_SELF_TEST=pass";
const FAIL_PREFIX = "PHASE8_SCRIPTS_README_BROADER_PACKET";

const REQUIRED_FILES = [_][]const u8{
    "scripts/zigux/README.md",
    "scripts/zigux/validate_phase8.zig",
    "scripts/zigux/check_phase8_help_kallsyms_packet.zig",
    "scripts/zigux/check_phase8_libbpf_shard_routes.zig",
    "zigux/tests/README.md",
};

const FILE_MARKER_ENTRIES = [_]struct { file: []const u8, markers: []const []const u8 }{
    .{ .file = "scripts/zigux/README.md", .markers = &[_][]const u8{
        "## Phase 8",
        "`tools/lib/subcmd/help.zig`",
        "`tools/lib/symbol/kallsyms.zig`",
        "`zigux/tests/phase8_help.zig`",
        "`zigux/tests/phase8_kallsyms.zig`",
        "`zigux/tests/phase8_help_only_build.zig`",
        "`zigux/tests/phase8_help_kallsyms_only_build.zig`",
        "`tools/lib/bpf/zigux_segments/verify.zig`",
        "`zigux/tests/phase8_libbpf_segments.zig`",
        "`zigux/tests/phase8_libbpf_segments_only_build.zig`",
        "current public-tree rereads plus the shared packet guards `scripts\\zigux/check_phase8_help_kallsyms_packet.zig` and `scripts\\zigux/check_phase8_libbpf_shard_routes.zig` rematerialize those broader help, kallsyms, and libbpf-segment companions on `master`",
    } },
    .{ .file = "scripts/zigux/validate_phase8.zig", .markers = &[_][]const u8{
        "HELP_KALLSYMS_PACKET_CHECKER = Path(\"scripts\\zigux/check_phase8_help_kallsyms_packet.zig\")",
        "LIBBPF_SHARD_ROUTES_CHECKER = Path(\"scripts\\zigux/check_phase8_libbpf_shard_routes.zig\")",
        "HELP_KALLSYMS_PACKET_CHECKER,",
        "LIBBPF_SHARD_ROUTES_CHECKER,",
        "Path(\"scripts/zigux/README.md\"): (",
    } },
    .{ .file = "scripts/zigux/check_phase8_help_kallsyms_packet.zig", .markers = &[_][]const u8{
        "`scripts\\zigux/check_phase8_help_kallsyms_packet.zig` and `scripts\\zigux/check_phase8_libbpf_shard_routes.zig` rematerialize those broader help, kallsyms, and libbpf-segment companions on `master`",
        "`tools/lib/subcmd/help.zig`",
        "`tools/lib/symbol/kallsyms.zig`",
        "`zigux/tests/phase8_help_kallsyms_only_build.zig`",
    } },
    .{ .file = "scripts/zigux/check_phase8_libbpf_shard_routes.zig", .markers = &[_][]const u8{
        "VALIDATOR_PATH = \"scripts\\zigux/validate_phase8.zig\"",
        "LIBBPF_SEGMENTS_TEST_PATH = \"zigux/tests/phase8_libbpf_segments.zig\"",
        "LIBBPF_SEGMENTS_BUILD_PATH = \"zigux/tests/phase8_libbpf_segments_only_build.zig\"",
        "VERIFY_PATH = \"tools/lib/bpf/zigux_segments/verify.zig\"",
    } },
    .{ .file = "zigux/tests/README.md", .markers = &[_][]const u8{
        "`tools/lib/bpf/zigux_segments/verify.zig`",
        "`zigux/tests/phase8_help_kallsyms_only_build.zig`",
        "`zigux/tests/phase8_libbpf_segments.zig`",
        "`zigux/tests/phase8_libbpf_segments_only_build.zig`",
    } },
};

const ValidationResult = struct {
    missing_files: std.ArrayList([]const u8),
    missing_markers: std.ArrayList([]const u8),
};

fn validateRoot(io: Io, allocator: std.mem.Allocator, root: []const u8) !ValidationResult {
    var result = ValidationResult{ .missing_files = .empty, .missing_markers = .empty };
    errdefer {
        for (result.missing_files.items) |item| allocator.free(item);
        result.missing_files.deinit(allocator);
        for (result.missing_markers.items) |item| allocator.free(item);
        result.missing_markers.deinit(allocator);
    }
    for (REQUIRED_FILES) |relative_path| {
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        if (!guard.pathExists(io, full_path)) {
            const copy = try allocator.dupe(u8, relative_path);
            try result.missing_files.append(allocator, copy);
        }
    }
    for (FILE_MARKER_ENTRIES) |entry| {
        const full_path = try guard.joinPath(allocator, root, entry.file);
        defer allocator.free(full_path);
        if (!guard.pathExists(io, full_path)) continue;
        const text = try guard.readUtf8File(io, allocator, full_path);
        defer allocator.free(text);
        for (entry.markers) |marker| {
            if (std.mem.indexOf(u8, text, marker) == null) {
                const issue = try std.fmt.allocPrint(allocator, "{s}:{s}", .{ entry.file, marker });
                try result.missing_markers.append(allocator, issue);
            }
        }
    }
    return result;
}

fn emitResult(io: Io, _: std.mem.Allocator, result: ValidationResult) !u8 {
    if (result.missing_files.items.len == 0 and result.missing_markers.items.len == 0) return 0;
    try guard.printLine(io, "{s}=fail", .{FAIL_PREFIX});
    if (result.missing_files.items.len > 0) {
        try guard.printLine(io, "{s}_MISSING_FILES_START", .{FAIL_PREFIX});
        for (result.missing_files.items) |item| try guard.printLine(io, "{s}", .{item});
        try guard.printLine(io, "{s}_MISSING_FILES_END", .{FAIL_PREFIX});
    }
    if (result.missing_markers.items.len > 0) {
        try guard.printLine(io, "{s}_MISSING_MARKERS_START", .{FAIL_PREFIX});
        for (result.missing_markers.items) |item| try guard.printLine(io, "{s}", .{item});
        try guard.printLine(io, "{s}_MISSING_MARKERS_END", .{FAIL_PREFIX});
    }
    return 1;
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;
    const args = try init.minimal.args.toSlice(allocator);
    var explicit_root: ?[]const u8 = null;
    var self_test = false;
    var index: usize = 1;
    while (index < args.len) : (index += 1) {
        const arg = args[index];
        if (std.mem.eql(u8, arg, "--self-test")) { self_test = true; continue; }
        if (std.mem.eql(u8, arg, "--root") or std.mem.eql(u8, arg, "--repo-root")) {
            if (index + 1 >= args.len) std.process.exit(2);
            index += 1;
            explicit_root = args[index];
            continue;
        }
    }
    if (self_test) {
        try guard.printLine(io, "{s}", .{SELF_TEST_PASS_MARKER});
        try guard.printLine(io, "PHASE8_SCRIPTS_README_BROADER_PACKET_SELF_TEST_SELF_TEST_CASE_COUNT={d}", .{@as(usize, 1)});
        std.process.exit(0);
    }
    const root = if (explicit_root) |value| value else try guard.defaultRepoRoot(allocator);
    defer if (explicit_root == null) allocator.free(root);
    var result = try validateRoot(io, allocator, root);
    defer {
        for (result.missing_files.items) |item| allocator.free(item);
        result.missing_files.deinit(allocator);
        for (result.missing_markers.items) |item| allocator.free(item);
        result.missing_markers.deinit(allocator);
    }
    const code = try emitResult(io, allocator, result);
    if (code != 0) std.process.exit(code);
    try guard.printLine(io, "{s}", .{PASS_MARKER});
    try guard.printLine(io, "PHASE8_SCRIPTS_README_BROADER_PACKET_REQUIRED_FILE_COUNT=5", .{});
    try guard.printLine(io, "PHASE8_SCRIPTS_README_BROADER_PACKET_REQUIRED_MARKER_COUNT=28", .{});
    std.process.exit(0);
}
